pkg/api/payer/nodeSelector.go

```diff
@@ -0,0 +1,61 @@
+package payer
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"github.com/xmtp/xmtpd/pkg/registry"
+	"github.com/xmtp/xmtpd/pkg/topic"
+	"sort"
+)
+
+type NodeSelectorAlgorithm interface {
+	GetNode(topic topic.Topic) (uint32, error)
+}
+
+type StableHashingNodeSelectorAlgorithm struct {
+	reg registry.NodeRegistry
+}
+
+func NewStableHashingNodeSelectorAlgorithm(
+	reg registry.NodeRegistry,
+) *StableHashingNodeSelectorAlgorithm {
+	return &StableHashingNodeSelectorAlgorithm{reg: reg}
+}
+
+// hashKey hashes the topic to a stable uint32 hash
+func HashKey(topic topic.Topic) uint32 {
+	hash := sha256.Sum256(topic.Bytes())
+	return binary.BigEndian.Uint32(hash[:4])
+}
+
+// GetNode selects a node for a given topic using stable hashing
+func (s *StableHashingNodeSelectorAlgorithm) GetNode(topic topic.Topic) (uint32, error) {
+	nodes, err := s.reg.GetNodes()
+	if err != nil {
+		return 0, err
+	}
+
+	if len(nodes) == 0 {
+		return 0, errors.New("no available nodes")
+	}
+
+	// Sort nodes to ensure stability
+	sort.Slice(nodes, func(i, j int) bool { return nodes[i].NodeID < nodes[j].NodeID })
+
+	topicHash := HashKey(topic)
+
+	// we do not want to hash the entire available hash space, just consider up to the largest NodeID issued
```

Comment:
I want to make sure my intuitions are correct here.

Let's say we have nodes 100, 200, and 300.

If we add node 400, what % of the topics will shift to point to a different node? I think by including the highest node ID in the `topicHash` we might defeat the purpose of consistent hashing, which is consistency of selection.

Am I misunderstanding this?

Reply:
Typically when I think of consistent hashing I think of having the hash be `% 360` (or some other constant) and distributing the node IDs equally between 0 and that constant.

Reply:
yes. I point out these limitations in the PR description. I am happy to brainstorm on how to get this right.

Reply:
Just spitballing

We define some constant for the size of the space. Let's say 360.

Then we split the nodes evenly over that space.

```python
spacing = 360 / num_nodes
node_locations = [spacing * index for index, node in enumerate(nodes)] // assumes nodes are sorted by ID
```

Then we can just `hash(topic) % 360` for each topic and put it somewhere in the space. As we add nodes the locations would move but only slightly.

Reply:
the new commit does that. Just in a 32bit space.